src/main/osd/osd_elements.c

```diff
@@ -2145,65 +2140,110 @@ static bool osdDrawSingleElementBackground(displayPort_t *osdDisplayPort, uint8_
-    element.buff = (char *)&buff;
-    element.osdDisplayPort = osdDisplayPort;
-    element.rendered = true;
-    element.drawElement = true;
+    activeElement.item = item;
+    activeElement.elemPosX = elemPosX;
+    activeElement.elemPosY = elemPosY;
+    activeElement.elemOffsetX = 0;
+    activeElement.elemOffsetY = 0;
+    activeElement.type = OSD_TYPE(osdElementConfig()->item_pos[item]);
+    activeElement.buff = elementBuff;
+    activeElement.osdDisplayPort = osdDisplayPort;
+    activeElement.rendered = true;
+    activeElement.drawElement = true;
 
     // Call the element background drawing function
-    osdElementBackgroundFunction[item](&element);
-    if (element.drawElement) {
-        osdDisplayWrite(&element, elemPosX, elemPosY, DISPLAYPORT_SEVERITY_NORMAL, buff);
+    osdElementBackgroundFunction[item](&activeElement);
+    if (activeElement.drawElement) {
+        displayPendingBackground = true;
     }
 
-    return element.rendered;
+    return activeElement.rendered;
 }
 
-static uint8_t activeElement = 0;
+static uint8_t activeElementNumber = 0;
+
+bool osdIsRenderPending(void)
+{
+    return displayPendingForeground | displayPendingBackground;
+}
 
 uint8_t osdGetActiveElement(void)
 {
-    return activeElement;
+    return activeElementNumber;
 }
 
 uint8_t osdGetActiveElementCount(void)
 {
     return activeOsdElementCount;
 }
 
+// Return true if there is more to display
+bool osdDisplayActiveElement(void)
+{
+    if (activeElementNumber >= activeOsdElementCount) {
+        return false;
+    }
+
+    // If there's a previously drawn background string to be displayed, do that
+    if (displayPendingBackground) {
+        osdDisplayWrite(&activeElement,
+                        activeElement.elemPosX + activeElement.elemOffsetX,
+                        activeElement.elemPosY + activeElement.elemOffsetY,
+                        activeElement.attr, activeElement.buff);
+
+        activeElement.buff[0] ='\0';
+
+        displayPendingBackground = false;
+
+        return displayPendingForeground;
+    }
+
+    // If there's a previously drawn foreground string to be displayed, do that
+    if (displayPendingForeground) {
+        osdDisplayWrite(&activeElement,
+                        activeElement.elemPosX + activeElement.elemOffsetX,
+                        activeElement.elemPosY + activeElement.elemOffsetY,
+                        activeElement.attr, activeElement.buff);
+
+        activeElement.buff[0] ='\0';
+
+        displayPendingForeground = false;
+    }
+
+    return false;
+}
+
 // Return true if there are more elements to draw
-bool osdDrawNextActiveElement(displayPort_t *osdDisplayPort, timeUs_t currentTimeUs)
+bool osdDrawNextActiveElement(displayPort_t *osdDisplayPort)
 {
-    UNUSED(currentTimeUs);
     static bool backgroundRendered = false;
     bool retval = true;
 
-    if (activeElement >= activeOsdElementCount) {
+    if (activeElementNumber >= activeOsdElementCount) {
```

Comment:
IMO this may kill OSD when switching OSD profile (new profile with smaller activeOsdElementCount)

Reply:
There will be no OSD update when `activeElementNumber >= activeOsdElementCount`. And there in IMO no code to reset `activeElementNumber`. But if OSD switch occurs only after full OSD update, this won't happen.

Reply:
any further changes needed, or is it  resolved with 5a979eca1ae9437e436b2cdae423005609b57410 as well?

Reply:
Good spot. Fixed.